### PR TITLE
feat: [Backend] Data Quality Monitoring with native compute, multi-backend support, REST API, CLI

### LINKED ISSUE #5919: Revamp Data Quality Monitoring
**Is your feature request related to a problem? Please describe.**

Feast defines and serves features via Feature Views, but it doesn’t provide a first-class way to monitor feature data quality and feature drift over time — especially for features produced via On Demand Transformations (ODT). 

Today, users often need to compare the “training distribution” (from a one-off static training dataset) with the distribution of feature values observed in production over time. This typically requires:

- custom online feature logging pipelines,
- bespoke Spark/SQL jobs to compute metrics and drift,
- and external dashboards/alerting systems.

This becomes even more complex when ODT is involved, since transformed feature values are not easily reproducible unless transformation provenance and feature versions are tracked.

**Describe the solution you'd like**

Introduce a new first-class concept in Feast: `DQMJob` (Data Quality Monitoring Job), which computes feature quality and drift metrics for one or more Feature Views over time and reports results in the Feast UI.

Proposed API:

```python
DQMJob(
    feature_view_names: List[str],
    data_source=None,  # should be the sink of the logged feature view values+metadata
    features_to_exclude: Optional[List[str]] = None,
    features_to_include: Optional[List[str]] = None,  # one or the other
    time_interval: Literal["day", "week", "month"] = "day",
    feature_drift_metrics_config=None,  # schema-based defaults + per-feature overrides
    baseline_config=None,  # static training dataset distribution reference
    logging_config=None,   # simple: 100% or sampled logging
)
````

Key components:

1. **Simple feature logging to an offline sink**

* During online feature retrieval, Feast can optionally emit feature logs to a configured offline sink (e.g., S3/GCS/ADLS or a warehouse table).
* Logging supports either:

  * 100% of requests, or
  * sampled (e.g., p=0.01).
* The sink is append-only and partitioned by time (e.g., dt=YYYY-MM-DD).
* This offline dataset represents the “live” production feature values used by DQMJob.

2. **ODT reproducibility via registry versioning**

* For On Demand Transformations, feature logs include:

  * input feature references + versions,
  * output feature references + versions,
  * transformation identifier (content hash),
  * registry revision (or equivalent version id).
* Transformation inputs/outputs and transformation code/artifact references are versioned in the Feast registry.
* A content-addressed hash defines transformation identity and derived feature versions, enabling deterministic reproduction.

3. **Baseline comparison against static training dataset**

* Baseline is the distribution from a one-off training dataset run (static).
* Each time bucket (e.g., daily) is compared against this fixed baseline using drift metrics appropriate to feature type.
* Baseline is referenced deterministically (e.g., `baseline_dataset_uri` or a registry-managed `baseline_snapshot_id`).

4. **Schema-aware metrics**

* Quality metrics (examples):

  * missing rate
  * min/max/range
  * mean/median
  * unique count / cardinality
* Drift metrics selected by schema:

  * numeric: PSI, KS-statistic, etc.
  * categorical: PSI, JS-divergence, entropy shift, etc.
* Defaults inferred from feature schema with per-feature overrides.

5. **Execution model**

* `DQMJob` compiles into OfflineStore-specific execution plans (Spark DataFrame operations, warehouse SQL, etc.).
* Jobs are intended to be run as batch jobs via orchestrators (Airflow, KFP, CLI, scripts).
* OfflineStore implementations expose extensions for computing metrics efficiently.

6. **Registry-managed metrics storage + Feast UI**

* Computed metrics are written to a registry-managed metrics table (or registry-owned storage location).
* Each metric record includes:

  * feature_view_name
  * feature_name
  * feature_version / registry_revision
  * metric_Name
  * metric_value
  * time_bucket
  * baseline_reference
* Feast UI queries this table to display:
  * metric time series
  * drift summaries
  * per-feature quality views

**Describe alternatives you've considered**

* Integrating Feast with third-party observability tools (Evidently, WhyLabs, Arize, etc.) using custom logging pipelines.
* Building custom Spark/SQL jobs on offline training data and online logs.
* Monitoring only offline data (which misses what is actually served online, especially with ODT).

While workable, these approaches require significant glue code and lack consistent offline/online symmetry within Feast.

**Additional context**

* Logging should be configurable for privacy and cost (sampling vs 100%).
* Feature logs should be stored in batch-friendly formats (e.g., Parquet/Delta) and partitioned by time.
* The design should preserve offline/online symmetry and leverage Feast registry versioning for reproducibility.
* The system should remain backend-agnostic while using OfflineStore extensions for computation.

### PR DESCRIPTION
**Pull request by jyejare**
# To check real UI monitoring:

Visit PR https://github.com/feast-dev/feast/pull/6422, see Demo.

# What this PR does / why we need it:

This PR introduces comprehensive feature quality monitoring capabilities to Feast, enabling proactive tracking of feature distributions and data quality metrics. Currently, Feast has no built-in tools for monitoring feature health in production — ML teams must build custom solutions to detect issues like distribution shifts, elevated null rates, or degraded data quality before they silently impact model performance.

**What it adds:**

### Core Monitoring Engine

- **Hybrid computation engine** — SQL push-down on the native `OfflineStore` as the primary compute path, with a Python-based (PyArrow/NumPy) fallback for backends that don't implement native compute. This leverages the offline store as a compute engine (same architecture as Feast materialization).
- **Fully native storage** — Monitoring metrics are stored within the configured `OfflineStore` backend itself (no separate monitoring database). Six static methods on the `OfflineStore` base class (`compute_monitoring_metrics`, `get_monitoring_max_timestamp`, `ensure_monitoring_tables`, `save_monitoring_metrics`, `query_monitoring_metrics`, `clear_monitoring_baseline`) handle compute and storage.
- **PyArrow-based metrics computation** (`MetricsCalculator`) — Backend-agnostic statistical computation as fallback, supporting:
  - Numeric features: mean, stddev, min/max, percentiles (p50/p75/p90/p95/p99), null rate, histograms
  - Categorical features: top-N value counts with other/unique counts
  - Automatic feature type classification from Feast's `PrimitiveFeastType` and `ValueType`

### Multi-Backend Support (8 Offline Stores)

All 6 native monitoring methods implemented for each backend with dialect-specific SQL:

| Backend | Compute | Storage | Dialect highlights |
|---------|---------|---------|-------------------|
| **PostgreSQL** | SQL push-down | `INSERT ON CONFLICT` | `PERCENTILE_CONT`, `WIDTH_BUCKET` |
| **Snowflake** | SQL push-down | `MERGE` with `VARIANT` JSON | `APPROX_PERCENTILE`, `WIDTH_BUCKET` |
| **BigQuery** | SQL push-down | `MERGE` into BQ tables | `APPROX_QUANTILES`, parameterized queries |
| **Redshift** | SQL push-down | `MERGE` via Data API | `APPROXIMATE PERCENTILE_DISC` |
| **Spark** | SparkSQL push-down | Parquet tables | `PERCENTILE_APPROX`, `spark.sql()` |
| **Oracle** | SQL via Ibis | `MERGE FROM DUAL` | `PERCENTILE_CONT WITHIN GROUP` |
| **DuckDB** | In-memory SQL | Parquet files | `QUANTILE_CONT`, `HISTOGRAM` |
| **Dask** | PyArrow compute | Parquet files | `pyarrow.compute` + `numpy` |

### Multi-Granularity Time-Series Metrics

- **5 granularities**: `daily`, `weekly`, `biweekly`, `monthly`, `quarterly`
- **Auto-compute mode**: Detects latest event timestamp and computes all granularities in one job
- **Pre-computed metrics** stored per date + granularity for fast retrieval
- **On-demand transient compute**: Fresh statistics for arbitrary date ranges (not stored)

### Batch + Log Data Source Support

- **Batch source**: Reads from the feature view's `batch_source` via `OfflineStore.pull_all_from_table_or_query()`
- **Log source**: Reads from feature serving logs via `FeatureService.logging_config` destination, using `__log_timestamp` as event timestamp
- **Feature name normalization**: Prefixed log column names (`driver_stats__conv_rate`) are parsed back to their original `feature_view_name` + `feature_name` for storage compatibility and drift detection
- **`data_source_type`** column (`batch` / `log`) differentiates metrics in storage

### Orchestration Service (`MonitoringService`)

- Ties registry, offline store, calculator, and storage together
- Computes and aggregates metrics at **feature**, **feature view**, and **feature service** levels
- Cached `OfflineStore` instance for performance
- Unified compute/timestamp methods handling both batch and log paths with SQL push-down + fallback

### NaN/Inf Sanitization

- Multi-layered protection against `NaN`/`Inf` float values that break JSON serialization:
  - `opt_float()` in `monitoring_utils.py` — sanitizes at SQL result parsing
  - `_sanitize_floats()` in `monitoring_service.py` — final safety net on all API read paths
  - Applied in PyArrow compute paths (MetricsCalculator, Dask backend)
- Prevents HTTP 422 errors from `Out of range float values are not JSON compliant: nan`

### Shared Utilities (`monitoring_utils.py`)

- Centralized table name constants, column lists, PK definitions
- `monitoring_table_meta()`, `opt_float()`, `empty_numeric_metric()`, `empty_categorical_metric()`, `normalize_monitoring_row()`, `build_view_aggregate()`
- Used by all 8 backends — eliminates duplication and ensures consistency

### DQM Job Engine (`DQMJobManager`)

- Asynchronous job abstraction for metric computation (`compute`, `baseline`, `auto_compute`)
- Job status tracking in `feast_monitoring_jobs` table
- Forwards all parameters including `set_baseline` to the compute engine
- Supports future integration with Ray/Spark job runners

### REST API (`/monitoring/`)

| Method | Endpoint | Description |
|--------|----------|-------------|
| `POST` | `/monitoring/compute` | Submit batch DQM job |
| `POST` | `/monitoring/auto_compute` | Auto-detect dates, all granularities |
| `POST` | `/monitoring/compute/transient` | On-demand compute (not stored) |
| `POST` | `/monitoring/compute/log` | Compute from serving logs |
| `POST` | `/monitoring/auto_compute/log` | Auto-detect log dates, all granularities |
| `GET` | `/monitoring/jobs/{job_id}` | DQM job status |
| `GET` | `/monitoring/metrics/features` | Per-feature metrics |
| `GET` | `/monitoring/metrics/feature_views` | Per-view aggregates |
| `GET` | `/monitoring/metrics/feature_services` | Per-service aggregates |
| `GET` | `/monitoring/metrics/baseline` | Baseline distribution retrieval |
| `GET` | `/monitoring/metrics/timeseries` | Time-series data for trend analysis |

All endpoints support cascading filters: `project`, `feature_service_name`, `feature_view_name`, `feature_name`, `granularity`, `data_source_type`, date range.

RBAC enforced using existing `AuthzedAction.DESCRIBE` (read) and `AuthzedAction.UPDATE` (compute).

### CLI (`feast monitor run`)

```
Options:
  --feature-view TEXT       Feature view name (omit for all)
  --feature-service TEXT    Feature service name (required for --source-type log with explicit dates)
  --feature-name TEXT       Feature name(s), repeatable
  --start-date TEXT         Start date YYYY-MM-DD (omit for auto-detect)
  --end-date TEXT           End date YYYY-MM-DD (omit for auto-detect)
  --granularity TEXT        daily | weekly | biweekly | monthly | quarterly
  --set-baseline            Mark this computation as baseline
  --source-type TEXT        batch | log | all (default: batch)
```

### Auto-Baseline on `feast apply`

- Automatically queues baseline metric computation for new features on `feast apply`
- Non-blocking (async DQM job), idempotent (skips existing baselines)
- **Configurable** — can be disabled via `feature_store.yaml`:

```yaml
data_quality_monitoring:
  auto_baseline: false
```

### Feast Operator Support

- New CRD type: `DataQualityMonitoringConfig` added to `FeatureStoreSpec`
- Operator generates `data_quality_monitoring` section in `feature_store.yaml` when config is set
- DeepCopy methods auto-generated via `make generate`
- Disabling auto-baseline from operator CR:

```yaml
apiVersion: feast.dev/v1
kind: FeatureStore
spec:
  feastProject: my_project
  dataQualityMonitoring:
    autoBaseline: false
```

### Documentation

- **How-to guide**: `docs/how-to-guides/feature-monitoring.md` — Production setup, CLI usage, REST API reference, orchestrator integration (Airflow, KFP, cron, K8s CronJob), backend compatibility table
- **Quickstart notebook**: `examples/monitoring/monitoring-quickstart.ipynb` — 12-step hands-on walkthrough with visualization examples
- `docs/SUMMARY.md` updated with links to both

**Design decisions:**

- **Native `OfflineStore` compute + storage** — Each backend implements its own SQL push-down for metrics calculation and uses its native UPSERT/MERGE for storage. No separate monitoring database needed.
- **Hybrid fallback** — Backends that don't implement native compute fall back to Python/PyArrow, ensuring all offline stores are supported.
- **Separate `/monitoring/` route** rather than extending existing `/metrics/` — The existing metrics route serves registry inventory metadata; monitoring serves statistical feature quality data with a different data path.
- **DQM Job Engine** for async computation — Supports future Ray/Spark integration for distributed metric computation.
- **Top-level `data_quality_monitoring` config** — Sits alongside `materialization` and `openlineage` in `RepoConfig`, reflecting that it spans offline store compute/storage + apply trigger + server API.

# Which issue(s) this PR fixes:

Partially Fixes #5919

# Checks

- [x] I've made sure the tests are passing.
- [x] My commits are signed off (`git commit -s`)
- [x] My PR title follows conventional commits format

## Testing Strategy

* Unit tests
* Integration tests
* Operator unit tests (Ginkgo)

**Test coverage (all passing):**

| Test Suite | Count | Covers |
|---|---|---|
| `test_metrics_calculator.py` | 30+ | Numeric/categorical computation, edge cases (empty, all-null, single value, high cardinality), type classification, PyArrow type classification, NaN/Inf sanitization, JSON serializability |
| `test_compute_correctness.py` | 40+ | Per-backend metric accuracy for all 8 offline stores (DuckDB, Dask, PostgreSQL, Snowflake, BigQuery, Redshift, Oracle, Spark) — numeric stats, histograms, categorical counts |
| `test_monitoring_integration.py` | 16+ | End-to-end batch/log computation, baseline flow, view/service aggregation, native storage dispatch, log feature name normalization, REST API endpoints, CLI, RBAC enforcement |
| `repo_config_test.go` | 92 | Operator repo config generation including DataQualityMonitoring config with auto_baseline disabled, YAML serialization verification |

**Snyk SAST scan:** 0 vulnerabilities across all new files.

### CODE REVIEW

File: sdk/python/feast/api/registry/rest/monitoring.py [hunk 1]
```diff
@@ -0,0 +1,371 @@
...
+
+
+def get_monitoring_router(grpc_handler, server=None):
+    router = APIRouter()
+
+    def _get_monitoring_service():
+        from feast.monitoring.monitoring_service import MonitoringService
+
+        store = server.store if server else grpc_handler.store
+        return MonitoringService(store)
```
Comment (ntkathole) [inline sdk/python/feast/api/registry/rest/monitoring.py]: This is recreated on every REST request. Either make `MonitoringService` a singleton per server, or lift the "tables ensured" check out to module/process level

File: sdk/python/feast/monitoring/monitoring_utils.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    """Compute view-level aggregate stats from per-feature metrics.
+
+    Returns a dict with keys: total_row_count, total_features,
+    features_with_nulls, avg_null_rate, max_null_rate.
+    """
+    null_rates = [
+        m["null_rate"] for m in metrics_list if m.get("null_rate") is not None
+    ]
+    return {
+        "total_row_count": metrics_list[0]["row_count"] if metrics_list else 0,
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/monitoring_utils.py]: instead of relying on metrics_list[0], which is brittle, do same as null_rates `max(m["row_count"] for m in metrics_list)`

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 3]
```diff
@@ -463,6 +474,627 @@ def offline_write_batch(
...
+    for col in columns:
+        p = f"p{merge_idx}"
+        merge_idx += 1
+        val = row.get(col)
+        if col == "histogram" and val is not None and not isinstance(val, str):
+            val = json.dumps(val)
+        param_type = _bq_scalar_param_type(col)
+        params.append(bigquery.ScalarQueryParameter(p, param_type, val))
+        using_parts.append(f"@{p} AS {col}")
+        on_parts.append(f"T.{col} = S.{col}")
```
[the remark this answers is not in the document]
Comment (ntkathole) [reply sdk/python/feast/infra/offline_stores/bigquery.py]: @jyejare this ^

File: sdk/python/feast/monitoring/dqm_job_manager.py [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+_JOBS_TABLE = "feast_monitoring_jobs"
+
+JOB_STATUS_PENDING = "pending"
+JOB_STATUS_RUNNING = "running"
+JOB_STATUS_COMPLETED = "completed"
+JOB_STATUS_FAILED = "failed"
+
+
+class DQMJobManager:
+    def __init__(self, config: PostgreSQLConfig):
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/dqm_job_manager.py]: so only supported for PostgreSQL ?

File: sdk/python/feast/monitoring/monitoring_service.py [hunk 5]
```diff
@@ -0,0 +1,1223 @@
...
+                        granularity=granularity,
+                        set_baseline=False,
+                        now=now,
+                    )
+                    total_features += len(metrics_list)
+                    granularities_computed.add(granularity)
+
+                self._compute_feature_service_metrics(
+                    project=project,
+                    granularity="daily",
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/monitoring_service.py]: why feature service only saves daily ? probably this block needs inside inside the granularity loop ?

File: sdk/python/feast/monitoring/monitoring_store.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    "computed_at",
+    "is_baseline",
+    "total_feature_views",
+    "total_features",
+    "avg_null_rate",
+    "max_null_rate",
+]
+
+
+class MonitoringStore:
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/monitoring_store.py]: what is this class for? don't see imported anywhere

File: infra/feast-operator/api/v1/featurestore_types.go [hunk 7]
```diff
@@ -127,6 +127,9 @@ type FeatureStoreSpec struct {
 	AuthzConfig     *AuthzConfig          `json:"authz,omitempty"`
 	CronJob         *FeastCronJob         `json:"cronJob,omitempty"`
 	BatchEngine     *BatchEngineConfig    `json:"batchEngine,omitempty"`
+	// Dqm configures Data Quality Monitoring behaviour.
+	// +optional
+	Dqm *DqmConfig `json:"dqm,omitempty"`
```
Comment (ntkathole) [inline infra/feast-operator/api/v1/featurestore_types.go]: dqm is not user friendly, I think it's better to name it `feature_monitoring` or something else

File: sdk/python/feast/infra/feature_servers/base_config.py [hunk 8]
```diff
@@ -83,6 +83,22 @@ class MetricsConfig(FeastConfigBaseModel):
     (feast_feature_freshness_seconds)."""
 
 
+class DqmInitialDistributionConfig(FeastConfigBaseModel):
```
Comment (ntkathole) [inline sdk/python/feast/infra/feature_servers/base_config.py]: I think these configs should also live at top instead of feature_server. It uses the *offline* store, not the online server. This field is more similar to `materialization`, which is top-level config. 

```
#feature_store.yaml        

feature_monitoring:   
  auto_baseline: false 
```

This matches the pattern: materialization: spans offline+online stores, openlineage: spans apply+materialize - feature_monitoring: spans offline store (compute/storage) + apply trigger + server API.

File: sdk/python/feast/monitoring/dqm_job_manager.py (hunk 4 above)
Comment (jyejare) [reply sdk/python/feast/monitoring/dqm_job_manager.py]: Great Catch 👍

File: sdk/python/feast/monitoring/monitoring_service.py (hunk 5 above)
Comment (jyejare) [reply sdk/python/feast/monitoring/monitoring_service.py]: Again great catch 👍

File: sdk/python/feast/monitoring/monitoring_store.py (hunk 6 above)
Comment (jyejare) [reply sdk/python/feast/monitoring/monitoring_store.py]: ahh its a dead code, This is the old approach from earlier in our implementation, a standalone PostgreSQL-hardcoded storage class for monitoring metrics. It was superseded when I moved to the fully native OfflineStore approach. Removed the code now.

File: infra/feast-operator/api/v1/featurestore_types.go (hunk 7 above)
Comment (jyejare) [reply infra/feast-operator/api/v1/featurestore_types.go]: @ntkathole dqm is very known term in developers, data scientists. Also feature_monitoring is a global scope term which covers both operational metrics + data quality metrics. So I keep it limited to data quality.

I can rephrase it to DataQualityMonitoring if you think thats better ?

File: sdk/python/feast/infra/feature_servers/base_config.py (hunk 8 above)
Comment (jyejare) [reply sdk/python/feast/infra/feature_servers/base_config.py]: This could be due to the existing metrics config is residing under server config. The metrics(even operational metrics) could also be computed for offline store. So I think its good if we move both operational metrics and dqm metrics under the parent monitoring config.

File: infra/feast-operator/api/v1/featurestore_types.go (hunk 7 above)
Comment (ntkathole) [reply infra/feast-operator/api/v1/featurestore_types.go]: yea, DataQualityMonitoring explicit is better than short name

**Comment by Srihari1192** [comment]
@jyejare Please find the few observation during testing 

1. The per-feature metrics query endpoint (GET /monitoring/metrics/features) rejects granularity=daily with HTTP 422 despite daily metrics being actively computed and stored by both POST /monitoring/auto_compute and feast monitor run -g daily — the data exists in storage but is unreachable via the query API due to daily being absent from the endpoint's granularity enum validator, while the CLI explicitly lists it as a valid option (-g [daily|weekly|biweekly|monthly|quarterly]).

``` for eg : repoducing steps failing for daily where as it successfull for weekly

    curl -k -s \
  '..../api/v1/monitoring/metrics/features?project=my_project&feature_view_name=driver_hourly_stats&granularity=daily'
{"status_code":422,"detail":"Out of range float values are not JSON compliant: nan","error_type":"ValueError"}%    
```


2. The monitoring service has incomplete dtype support across both compute and read paths — Map/Json/Struct features are silently skipped at compute time with WARNING: Unsupported dtype log lines and never stored, while String features are computed and stored correctly but cause HTTP 422 on GET /monitoring/metrics/features due to a response schema mismatch between categorical metrics format (top-N histogram) and the numeric-only response model. In both cases the job reports Status: completed with total_features reflecting only successfully computed features rather than total registered features, giving no indication that any features were skipped or unreadable.
    
error logs with feast cli  ```Auto-computing batch metrics for all granularities...
WARNING:feast.monitoring.monitoring_service:Unsupported dtype 'Map' for feature 'driver_metadata', skipping
WARNING:feast.monitoring.monitoring_service:Unsupported dtype 'Json' for feature 'driver_config', skipping
WARNING:feast.monitoring.monitoring_service:Unsupported dtype 'Struct({name: String, age: String})' for feature 'driver_profile', skipping
Status: completed
Feature views computed: 1
Features computed: 15
Granularities: biweekly, daily, monthly, quarterly, weekly
Duration: 293ms```

**Comment by jyejare** [comment]
@Srihari1192 Both the issues are fixed and regarding the other observation about Map/Json/Struct features being silently skipped — that's working as designed. These complex types don't have meaningful statistical metrics (mean, stddev, etc.). The WARNING log is the correct behavior, and computed_features correctly reflects how many features were actually computed. No change needed there.

File: sdk/python/feast/monitoring/dqm_job_manager.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+                    feature_view_name=job.get("feature_view_name"),
+                    feature_names=params.get("feature_names"),
+                    start_date=date.fromisoformat(params["start_date"])
+                    if params.get("start_date")
+                    else None,
+                    end_date=date.fromisoformat(params["end_date"])
+                    if params.get("end_date")
+                    else None,
+                    granularity=params.get("granularity", "daily"),
+                )
```
[the remark this answers is not in the document]
Comment (jyejare) [reply sdk/python/feast/monitoring/dqm_job_manager.py:140]: Resolved in latest commit.

File: sdk/python/feast/infra/offline_stores/bigquery.py [hunk 10]
```diff
@@ -477,6 +489,645 @@ def offline_write_batch(
...
+                f"APPROX_QUANTILES({c}, 100)[OFFSET(75)] AS c{i}_p75",
+                f"APPROX_QUANTILES({c}, 100)[OFFSET(90)] AS c{i}_p90",
+                f"APPROX_QUANTILES({c}, 100)[OFFSET(95)] AS c{i}_p95",
+                f"APPROX_QUANTILES({c}, 100)[OFFSET(99)] AS c{i}_p99",
+            ]
+        )
+    query = (
+        f"SELECT {', '.join(select_parts)} "
+        f"FROM {from_expression} AS _src WHERE {ts_filter}"
+    )
```
[the remark this answers is not in the document]
Comment (jyejare) [reply sdk/python/feast/infra/offline_stores/bigquery.py:982]: Resolved in latest commit.

**Comment by ntkathole** [comment]
@jyejare Please fix the CI

File: infra/feast-operator/config/crd/bases/feast.dev_featurestores.yaml [hunk 11]
```diff
@@ -529,6 +529,16 @@ spec:
                     description: The time zone name for the given schedule, see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones.
                     type: string
                 type: object
+              dqm:
```
Comment (ntkathole) [inline infra/feast-operator/config/crd/bases/feast.dev_featurestores.yaml]: @jyejare looks like only Go type name changed, this dqm is not renamed, which is user facing and important

File: sdk/python/feast/repo_config.py [hunk 12]
```diff
@@ -358,6 +365,9 @@ class RepoConfig(FeastBaseModel):
     mlflow_config: Optional[Any] = Field(None, alias="mlflow")
     """ MlflowConfig: Configuration for MLflow experiment tracking integration (optional). """
 
+    dqm_config: Optional[DqmConfig] = Field(None, alias="dqm")
```
Comment (ntkathole) [inline sdk/python/feast/repo_config.py]: same here as well, should be feature_monitoring or data_quality_monitoring

File: sdk/python/feast/monitoring/monitoring_service.py [hunk 13]
```diff
@@ -0,0 +1,1254 @@
...
+        if granularity not in VALID_GRANULARITIES:
+            raise ValueError(
+                f"Invalid granularity '{granularity}'. "
+                f"Must be one of {VALID_GRANULARITIES}"
+            )
+
+        start_time = time.time()
+        start_dt, end_dt = self._to_date_range(start_date, end_date)
+
+        if project is None:
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/monitoring_service.py]: project is required in definition, so none check here is not required

File: sdk/python/feast/cli/monitor.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+        )
+        click.echo(f"Status: {result['status']}")
+        click.echo(f"Feature services computed: {result['computed_feature_services']}")
+        click.echo(f"Features computed: {result['computed_features']}")
+        click.echo(f"Granularities: {', '.join(result['granularities'])}")
+        click.echo(f"Duration: {result['duration_ms']}ms")
+    else:
+        if not feature_service_name:
+            click.echo(
+                "Error: --feature-view (as feature service name) is required for log source with explicit dates."
```
Comment (ntkathole) [inline sdk/python/feast/cli/monitor.py]: isn't this confusing ? --feature-service option would be clearer

File: sdk/python/feast/monitoring/monitoring_service.py [hunk 15]
```diff
@@ -0,0 +1,1254 @@
...
+                    logger.warning(
+                        "No data found for feature view '%s', skipping", fv.name
+                    )
+                    continue
+
+                now = datetime.now(timezone.utc)
+
+                for granularity, window in GRANULARITY_WINDOWS.items():
+                    window_start = max_ts - window
+                    metrics_list = self._compute_feature_metrics(
```
Comment (ntkathole) [inline sdk/python/feast/monitoring/monitoring_service.py:123]: This fires 5 separate SQL pushdown queries against the same source table (one per granularity: daily/weekly/biweekly/monthly/quarterly). A single query over the full quarterly window would be more efficient.

Comment (jyejare) [reply sdk/python/feast/monitoring/monitoring_service.py:123]: Each granularity computes over a different time window (daily=1d, weekly=7d, etc.), producing different statistical results. A single query over the quarterly window would give incorrect daily/weekly/monthly metrics. This is an inherent design requirement, not an optimization opportunity.

File: sdk/python/feast/repo_operations.py [hunk 16]
```diff
@@ -1,594 +1,630 @@
...
+        # Cleanup is handled in the new _apply_diffs method
+        pass
+
+    _submit_baseline_jobs_if_needed(store, project_name, repo)
+
+
+def _submit_baseline_jobs_if_needed(store, project_name, repo):
+    """Submit async baseline DQM jobs for new features after feast apply."""
+    dqm = store.config.data_quality_monitoring_config
+    if dqm is not None and not dqm.auto_baseline:
```
Comment (ntkathole) [inline sdk/python/feast/repo_operations.py]:
```suggestion
    if dqm is None or not dqm.auto_baseline:
```

File: sdk/python/feast/repo_operations.py [hunk 17]
```diff
@@ -1,594 +1,630 @@
...
+        if not feature_views:
+            return
+
+        job_ids = svc.submit_baseline_for_new_features(
+            project=project_name, feature_views=feature_views
+        )
+        for job_id in job_ids:
+            click.echo(f"  → Queued baseline metrics computation (DQM job: {job_id})")
+            try:
+                svc.execute_job(job_id)
```
Comment (ntkathole) [inline sdk/python/feast/repo_operations.py]: For repos with many feature views and large data sources, this could make feast apply extremely slow. May be run async ?

File: sdk/python/feast/api/registry/rest/system_metrics.py [hunk 18]
```diff
@@ -0,0 +1,173 @@
+import logging
```
Comment (ntkathole) [inline sdk/python/feast/api/registry/rest/system_metrics.py]: Let's keep this separate

File: sdk/python/feast/infra/offline_stores/duckdb.py [hunk 19]
```diff
@@ -113,6 +132,336 @@ def _write_data_source(
...
 
+# ------------------------------------------------------------------ #
+#  DuckDB monitoring (Parquet-backed)
+# ------------------------------------------------------------------ #
+
+MONITORING_DIR = "feast_monitoring"
+FEATURE_METRICS_FILE = "feature_metrics.parquet"
+VIEW_METRICS_FILE = "feature_view_metrics.parquet"
+SERVICE_METRICS_FILE = "feature_service_metrics.parquet"
+JOB_METRICS_FILE = "jobs.parquet"
```
Comment (ntkathole) [inline sdk/python/feast/infra/offline_stores/duckdb.py]: This could be a single function in monitoring_utils.py like:

MONITORING_PARQUET_FILES = {
    "feature": "feature_metrics.parquet",
    "feature_view": "feature_view_metrics.parquet",
    "feature_service": "feature_service_metrics.parquet",
    "job": "jobs.parquet",
}
 Same repeated across duckdb and dask.
 
Also, inconsistency with how SQL backends name things. `monitoring_utils.py` already defines table names, But the Parquet files use different names (feature_metrics.parquet vs table name feast_monitoring_feature_metrics). This inconsistency means a _duckdb_mon_table_meta function is needed to map between them, when the names could just match.

File: sdk/python/feast/repo_operations.py (hunk 16 above)
Comment (jyejare) [reply sdk/python/feast/repo_operations.py]: I have move the block to apply function and it triggers baseline calculation when both dqm and auto_baseline is enabled. Thats the right place.

File: sdk/python/feast/repo_operations.py (hunk 17 above)
Comment (jyejare) [reply sdk/python/feast/repo_operations.py]: Fixed with multithreading.

**Comment by ntkathole** [comment]
@jyejare please resolve conflicts

**Review by ntkathole — APPROVED** [review]
Looks good
